clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -41,6 +42,150 @@ struct is_iterator<T *> : std::true_type {};
 
 #ifndef DPCT_USM_LEVEL_NONE
 
+// device_allocator_traits is a helper struct which makes use of custom
+//  allocator constructor routines when they are specified serially on the host,
+//  while enabling oneDPL pstl accelleration when such custom constructors are
+//  not provided.
+template <typename _Allocator> struct device_allocator_traits {
+
+  // taken from libc++
+  template <class, class _Alloc, class... _Args>
+  struct __has_construct_impl : ::std::false_type {};
+
+  template <class _Alloc, class... _Args>
+  struct __has_construct_impl<decltype((void)std::declval<_Alloc>().construct(
```

Comment:
I think we can use `__has_construct` directly here, like `__has_destroy`.

Reply:
I don't believe its possible to use directly because of the variable number of arguments `construct()` can accept.  The reason this works for `__has_destroy` gets to be simpler, is that we know that it only accepts a single pointer as an argument.  

They are effectively using the same trick, but because of the template variable pack, we have to utilize the additional template argument at the beginning and use the `_impl` layer rather than at the end of the template parameters with a defaulted argument.